zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2152,6 +2150,19 @@ public void connect(final long maxMsToWaitUntilConnected, Watcher watcher)
 
       IZkConnection zkConnection = getConnection();
       _eventThread = new ZkEventThread(zkConnection.getServers());
+
+      if (_monitor != null) {
+        _monitor.setAndInitZkEventThreadMonitor(_eventThread);
+      }
+
+      try {
```

Comment:
Note, It seems crucial to register here. Or depending on timing, we may lose read of the first sync().

Reply:
I debugged with your code, so the ZkClientPathMonitor objects are currently only created in the register() method. So that's the reason we lose the count if register later.

There is a simple fix, we can instantiate all the ZkClientPathMonitors in the ZkClientMonitor constructor. Like this,
    for (ZkClientPathMonitor.PredefinedPath path : 
```
ZkClientPathMonitor.PredefinedPath.values()) {
      // If monitor root path only, check if the current path is Root.
      // Otherwise, add monitors for every path.
      if (!_monitorRootOnly || path.equals(ZkClientPathMonitor.PredefinedPath.Root)) {
        _zkClientPathMonitorMap.put(path,
            new ZkClientPathMonitor(path, _monitorType, _monitorKey, _monitorInstanceName));
      }
    }
```
Then in the register(), just register all the ZkClientPathMonitor objects that are in the _zkClientPathMonitorMap.

The good thing is that it seems to be the right thing to do.

Reply:
implemeted this way